design/anonymousTokens.md

```diff
@@ -0,0 +1,96 @@
+
+## Background
+
+* When a Narrative docker container is spun up for a user, the narrative launcher expects a KBase
+  auth token and assigns the container to the user.
+* JSON workspace object DLs go through the Data Import Export (DIE) server, which appears
+  to handle anonymous requests correctly based on examination of the code (would need to be
+  tested).
+* Other DLs go through the Narrative Job Server (NJS), which requires a token - and all KBase
+  Software Development Kit (SDK) apps require a token - to create the DL package in Shock.
+  * DIE is responsible for DLing the package from Shock for the user in the case of
+    non-staging area DLs
+      * DIE appears to handle anonymous DLs correctly based on examination of the
+        code (would need testing).
+      * HOWEVER, the Shock node would need to be publicly readable, and no DL apps
+        make that change (because they all expect a token).
+
+## Document purpose
+
+Propose a preliminary design for KBase auth service support of anonymous tokens (ATs)
+that allow anonymous users to have a transient identity when interacting with KBase. Discuss how
+those tokens would interact with the rest of the KBase infrastructure and what changes would be
+necessary.
+
+## Anonymous tokens
+
+* The KBase auth service would be enhanced to offer AT support.
+* ATs would have the same fields as regular tokens except that the the `type` would
+  always be `Anonymous`.
+* ATs would have a random user name assigned upon creation.
+  * Could be a concatenated dictionary style name for mnemonic purposes
+    (e.g. [https://gfycat.com/shoddydismalamoeba](https://gfycat.com/shoddydismalamoeba)).
+      * Need to check collision frequency.
+* ATs would exist in a separate token and user namespace from standard tokens.
+* ATs could not be used to create other tokens.
+* ATs would be prefixed with a string to indicate they are anonymous
+  (e.g. `***ANONYMOUS***-[remainder of token goes here]`)
+* ATs would be created, introspected, and revoked with separate APIs from normal tokens.
+  * Token creation requires no authentication.
+* The administration UI would have AT and user lookup functionality and the ability
+  to revoke said tokens.
+* The administration UI would have the ability to set the lifetime of ATs. The default lifetime
+  could be fairly long since ATs grant few privileges.
+* All endpoints except for the AT specific endpoints would respond with an error when
+  presented with an AT.
+  
+## Interactions with KBase infrastructure
+
+* Most KBase core services (including data services) should never see an an AT.
+* Narrative:
+  * When a user without a KBase cookie in the browser loads the KBase UI, an anonymous token would
+    be generated and stored in a cookie (potentially the same cookie).
+  * Starting a narrative would assign the docker container to the username in the AT.
+    * Narratives started in this manner would always be read-only to the user and must
```

Comment:
Noting here that the Narrative Interface will need to be aware of anonymous tokens, so it can manage its user session properly (i.e., pretend that there isn't a token, don't pass it to the Workspace service to fetch data, etc.)

Reply:
except for NJS - need to send ATs to NJS to start DL jobs.

Reply:
Does `The Narrative backend would not pass the AT on to any services.` not cover this? The DL part later covers the 2nd comment, yeah?

Reply:
Yup. That covers it.